packages/@aws-cdk/aws-iam/lib/principals.ts

```diff
@@ -82,6 +82,77 @@ export abstract class PrincipalBase implements IPrincipal {
     // Have to implement toJSON() because the default will lead to infinite recursion.
     return this.policyFragment.principalJson;
   }
+
+  /**
+   * Returns a new PrincipalWithConditions using this principal as the base, with the
+   * passed conditions added.
+   *
+   * When there is a value for the same operator and key in both the principal and the
+   * conditions parameter, the value from the conditions parameter will be used.
+   *
+   * @returns a new PrincipalWithConditions object.
+   */
+  public withConditions(conditions: Conditions): IPrincipal {
+    return new PrincipalWithConditions(this, conditions);
+  }
+}
+
+/**
+ * An IAM principal with additional conditions specifying when the policy is in effect.
+ *
+ * For more information about conditions, see:
+ * https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition.html
+ */
+export class PrincipalWithConditions<PrincipalType extends PrincipalBase> implements IPrincipal {
+  public readonly conditions: Conditions;
+  public readonly grantPrincipal: IPrincipal = this;
+  public readonly assumeRoleAction: string = 'sts:AssumeRole';
+
+  constructor(
+    public readonly principal: PrincipalType,
+    conditions: Conditions,
+  ) {
+    // Copy the intiial conditions from the principal
+    this.conditions = JSON.parse(JSON.stringify(this.principal.policyFragment.conditions));
```

Comment:
This feels dangerous, I'm not sure tokens will survive this properly. What if we just keep `conditions` not as a mutable copy of the original, but as a list of mutations? So it starts out empty, and we deep-merge it upon rendering?

Reply:
I was umm-ing and ahh-ing a bit over the approach here. I went with the copy largely because it was simpler than having a deep merge happen on `principal.policyFragment` and `principal.conditions` (so that the `conditions` getter returns the same thing the final render will include), but I don't have a good sense of how tokens work. If it's likely to break tokens, I'm happy to make the switch.

Reply:
Can you please try that? It will make me feel better.